### ISSUE TITLE
$PIPX_HOME on macOS since release 1.3.0 breaks awscli

### ISSUE THREAD

**Issue by hhp21** [pypa/pipx#1198]
It appears that since release 1.3.0, when installing pipx on macOS Ventura, it chooses to set `$PIPX_HOME` to `~/Library/Application Support/pipx`, which creates a problem for scripts like `awscli` where the directory forms part of the shebang, because shebangs are not allowed to have spaces in it.

As recent as release 1.2.1, $PIPX_HOME on macOS Ventura is set to be `~/.local/pipx` which doesn't cause problems.

**How to reproduce**

I was using Python 3.8 installed using pyenv.
```
pip install pipx==1.3.0
pipx environment 
```
It prints the below:
<img width="506" alt="Screenshot 2024-01-08 at 18 43 56" src="https://github.com/pypa/pipx/assets/12730232/a30fe0e2-e53e-498a-a960-b5b9a4126a3d">

**Expected behavior**

`$PIPX_HOME` should be `/Users/test3/.local/pipx`

**Comment by Gitznik** [comment]
Since 1.3, pipx bases the `$PIPX_HOME` on [platformdirs](https://github.com/platformdirs/platformdirs?tab=readme-ov-file#example-output), see [the docs](https://pipx.pypa.io/stable/installation/#installation-options). This package specifies for mac, that the canonical application data dir is `~/Library/Application Support/{app_name}`. `/Users/{user}/.local` is the canonical path on Linux.

I personally think it makes sense to stick to the platform specific convention. 

For an immediate fix to your problem you could symlink the binary you want to use in a shebang to a directory without spaces in the path and use that in the shebang.

We should definitely add this as a warning on the docs for mac though IMO.

**Comment by hhp21** [comment]
Thanks for the explanation. I've updated the title to be more precise.

**Comment by Gitznik** [comment]
I found quite a lengthy discussion (https://github.com/platformdirs/platformdirs/issues/4) on the platformdirs repo about suporting [XDG](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), which is essentially what you're asking for.

They ended up not supporting it for macOS. There was an interesting comment though on symlinking `~/Library/Application Support` to `~/Library/ApplicationSupport`. This apparently is done already if you're a Iterm2 user, but IMO this is generally something we should suggest in our docs, when using pipx to install applications that should be used in a shebang.

@gaborbernat as maintainer of platformdirs, do you have any other insight here?

**Comment by gaborbernat** [comment]
Nothing besides what was already discussed in that issue 🤔

**Comment by Gitznik** [comment]
I would suggest adding a note to the mac installation docs, in regards to symlinking `~/Library/Application Support` to `~/Library/ApplicationSupport` then.

**Comment by gaborbernat** [comment]
Well this only happens for awscli, no?

**Comment by Gitznik** [comment]
Any application one would use in a shebang I guess. Not that I could think of another example to be fair.

**Comment by gaborbernat** [comment]
You are welcome to put in a PR to be added to some FAQ section.

**Comment by jaraco** [comment]
Based on [this comment](https://github.com/pypa/pipx/pull/1204#issuecomment-2018091781), I propose this issue should be re-opened.

**Comment by chrysle** [comment]
It is currently being worked on reverting the `platformdirs` usage for `pipx` paths on macOS in #1297, since several issues with it have been reported. On other Unix systems the behaviour will be kept.

Besides, probably this issue should be reported at the `platformdirs` location, too.

**Comment by jaraco** [comment]
I see Gitznik in [this comment](https://github.com/pypa/pipx/issues/1212#issuecomment-1910656091) proposed that the issue be solved with in pip or the packages themselves. Personally, I'd prefer to see the issue solved at the core and add support for spaces in shebangs. Let's stop exposing this broken behavior on downstream workarounds.

The symlink workaround is terrible and should be avoided.

**Comment by jaraco** [comment]
> I think maybe the reason it doesn't affect console scripts is because pip uses a different executable depending on the type of script installed:

I was mistaken. I thought I'd installed pipx using pipx, but I see on this machine I have pipx installed using homebrew. The reason console scripts aren't affected by the shebang in console scripts is because pip provides a workaround for those scripts:

```
 ~ @ head -n 2 $(which tox)
#!/bin/sh
'''exec' "/Users/jaraco/Library/Application Support/pipx/venvs/tox/bin/python" "$0" "$@"
```

**Comment by Gitznik** [comment]
> > On other Unix systems the behaviour will be kept.
> 
> That's too bad, because other Unix users are affected too, except that they avoid having spaces in their directory names and perpetuate the expectation that spaces are invalid.

This is about making sure the default pipx behavior is robust. We can't really prevent users from setting their pipx home to a path containing spaces. I may extend my PR to print a warning if a space is found in the pipx home.

I don't think this project is the right place to discuss whether spaces should or should not be allowed in shebangs.

**Comment by jaraco** [comment]
Today I learned that [the shell I use](https://xon.sh) actually supports spaces in pathnames in shebangs. Since it uses `shlex` to parse the shebang line, it works if one puts quotes around the interpreter. In https://github.com/xonsh/xonsh/discussions/5315, I provide more details and I'm seeking to find places to make a proper solution more widely available.

**Comment by frou** [comment]
I run `pipx environment --value PIPX_BIN_DIR` as part of my `.bashrc` (adding it to `PATH`), and since updating to pipx 1.5.0, that means there is stderr spam every time I start a shell.

```txt
⚠️ Found a space in the home path. We heavily discourage this, due to multiple
    incompatibilities. Please check our docs for more information on this, as well as
    some pointers on how to migrate to a different home path.
```

<img width="621" alt="Screenshot 2024-03-31 at 09 50 40" src="https://github.com/pypa/pipx/assets/172663/5b158074-8d5f-449e-b04b-fe9e48999ccc">

**Comment by Gitznik** [comment]
> I run `pipx environment --value PIPX_BIN_DIR` as part of my `.bashrc` (adding it to `PATH`), and since updating to pipx 1.5.0, that means there is stderr spam every time I start a shell.

Yes, this is intended to warn users that having a pipx_home path with spaces in it can lead to a lot of problems. See [this discussion](https://github.com/pypa/pipx/discussions/1247) for more details.

To solve this you can
1. Move your pipx home to a location without spaces (see our [troubleshooting docs](https://pipx.pypa.io/stable/troubleshooting/#pipx-files-not-in-expected-locations-according-to-documentation)) - I would heavily urge you to do this
2. If you're really sure you want to stick to your old path, you can run the command with the `-q` flag: `pipx environment -q --value PIPX_BIN_DIR`. This will silence all warnings.

**Comment by Gitznik** [comment]
I will actually close this issue, as with version 1.5, this is no longer an issue in the default pipx behavior.

**Comment by frou** [comment]
Thank you. So I just did `pipx uninstall-all`, then manually deleted the entire `~/Library/Application Support/pipx` directory, then `pipx install`'d the handful of stuff I use pipx for, afresh. And now it's good.

Where I was personally confused (and possibility others who aren't pipx experts will be too) is that I was reading in to this some kind of implication that I had make the choice to use `~/Library/Application Support/pipx`, when IIRC I never did.

**Comment by Gitznik** [comment]
> Thank you. So I just did `pipx uninstall-all`, then manually deleted the entire `~/Library/Application Support/pipx` directory, then `pipx install`'d the handful of stuff I use pipx for, afresh. And now it's good.
> 
> Where I was personally confused (and possibility others who aren't pipx experts will be too) is that I was reading in to this some kind of implication that I, as a user, had made the choice to use `~/Library/Application Support/pipx` (which IIRC wasn't the case).

Gotcha. If you have an idea how to word this more clearly, you're welcome to open a PR :) The warning is generated here:https://github.com/pypa/pipx/blob/8fbc0850e83d50f8c9d224da94b79dad55dca70e/src/pipx/paths.py#L114-L116
